gluten-core/src/main/scala/io/glutenproject/extension/CollapseProjectTransformer.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.extension
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.execution.ProjectExecTransformer
+
+import org.apache.spark.sql.catalyst.optimizer.CollapseProjectShim
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.SparkPlan
+
+object CollapseProjectTransformer extends Rule[SparkPlan] {
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    if (!GlutenConfig.getConf.enableColumnarProjectCollapse) {
+      return plan
+    }
+    plan.transformUp {
+      case p1 @ ProjectExecTransformer(_, p2: ProjectExecTransformer)
+          if CollapseProjectShim.canCollapseExpressions(
+            p1.projectList,
+            p2.projectList,
+            alwaysInline = false) =>
+        p2.copy(projectList =
+          CollapseProjectShim.buildCleanedProjectList(p1.projectList, p2.projectList))
```

Comment:
To be conservative, better to check if the collapsed project can be transformed, and fallback to original plan if fail to validate. Also, do not forget to copy tags.

Reply:
> do not forget to copy tags.

Do you mean that after executing `val newProject = p2.copy`, I still need to call `newProject.copyTagsFrom(p1)`?

Reply:
it seems there is no way to show the fallback reason for a transform node.. so it's ok to skip copying tags.